doc/tutorials/chapter_8/02_finite_state_machine.md

```diff
@@ -0,0 +1,353 @@
+          // OvenState will changed to `OvenState.completed` state.
+          events: {
+            Logic(name: 'button_pause')
+                  ..gets(button
+                      .eq(Const(Button.pause.value, width: button.width))):
+                OvenState.paused,
+            Logic(name: 'counter_time_complete')..gets(counter.val.eq(4)):
+                OvenState.completed
+          },
+          // actions:
+          // During the cooking state, `led` is change to yellow; timer's
+          // `counterReset` is set to 0 (Do not reset);
+          // timer's `en` is set to 1 (Enable value update).
+          actions: [
+            led < LEDLight.yellow.value,
+            counterReset < 0,
+            en < 1,
+          ]),
+
+      // identifier: paused state, represent by `OvenState.paused`.
+      State<OvenState>(OvenState.paused,
+          // events:
+          // When the button `resume` is pressed during paused state,
+          // OvenState will changed to `OvenState.cooking` state.
+          events: {
+            Logic(name: 'button_resume')
+                  ..gets(button
+                      .eq(Const(Button.resume.value, width: button.width))):
+                OvenState.cooking
+          },
+          // actions:
+          // During the paused state, `led` is change to red; timer's
+          // `counterReset` is set to 0 (Do not reset);
+          // timer's `en` is set to 0 (Disable value update).
+          actions: [
+            led < LEDLight.red.value,
+            counterReset < 0,
+            en < 0,
+          ]),
+
+      // identifier: completed state, represent by `OvenState.completed`.
+      State<OvenState>(OvenState.completed,
+          // events:
+          // When the button `start` is pressed during completed state,
+          // OvenState will changed to `OvenState.standby` state.
+          events: {
+            Logic(name: 'button_start')
+                  ..gets(button
+                      .eq(Const(Button.start.value, width: button.width))):
+                OvenState.standby
+          },
+          // actions:
+          // During the start state, `led` is change to green; timer's
+          // `counterReset` is set to 1 (Reset value);
+          // timer's `en` is set to 0 (Disable value update).
+          actions: [
+            led < LEDLight.green.value,
+            counterReset < 1,
+            en < 0,
+          ])
+    ];
+```
+
+By now, you already have a list of `state` ready to be passed to the `StateMachine`. Let assign the the `state` to the StateMachine declared. Note that, we also passed `OvenState.standby` to the StateMachine to understand that is the State when reset signal is given.
+
+```dart
+_oven = StateMachine<OvenState>(clk, reset, OvenState.standby, states);
+```
+
+Let also create a getter to expose `_oven` to the public as we can used it to generate a FSM diagram.
+
+```dart
+StateMachine<OvenState> get ovenStateMachine => _oven;
```

Comment:
This is maybe a dangerous practice to export this from a module hierarchy because it gives access to pins inside (violation of input/output module rules if used)